File: online/online.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>

using namespace std;

struct Bug {
    int edgeId;
    int bugId;
    int parentBugId;
    int occurrences;
    double percentageContribution;

    Bug(int edgeId, int bugId, int parentBugId, int occurrences, double percentageContribution)
        : edgeId(edgeId), bugId(bugId), parentBugId(parentBugId), occurrences(occurrences), percentageContribution(percentageContribution) {}

    bool isLeaf() {
        return occurrences != -1;
    }
};

vector<Bug> readBugsFromFile(const string& filename) {
    vector<Bug> bugs;
    ifstream file(filename);
    if (file.is_open()) {
        string line;
        while (getline(file, line)) {
            if (line.find("edge_id") != string::npos) {
                continue;
            }

            istringstream iss(line);
            string token;
            vector<string> parts;
            while (getline(iss, token, '\t')) {
                parts.push_back(token);
            }

            int edgeId = stoi(parts[0]);
            int bugId = stoi(parts[1].substr(3)); // Remove "bug" prefix
            int parentBugId = (parts[2] == "null") ? -1 : stoi(parts[2].substr(3)); // Remove "bug" prefix
            int occurrences = (parts[3] == "null") ? -1 : stoi(parts[3]);
            double percentageContribution = (parts[4] == "null") ? 0.0 : stod(parts[4]);

            bugs.push_back(Bug(edgeId, bugId, parentBugId, occurrences, percentageContribution));
        }
        file.close();
    }
    return bugs;
}

void calculateOccurrences(const vector<Bug>& bugs, Bug& bug, map<int, int>& bugOccurrences) {
    if (bugOccurrences.count(bug.bugId)) {
        return;
    }

    if (bug.isLeaf()) {
        bugOccurrences[bug.bugId] = bug.occurrences;
        return;
    }

    int totalOccurrences = 0;
    for (const Bug& child : bugs) {
        if (child.parentBugId == bug.bugId) {
            calculateOccurrences(bugs, child, bugOccurrences);
            int childOccurrences = bugOccurrences[child.bugId];
            totalOccurrences += (childOccurrences * child.percentageContribution) / 100.0;
        }
    }
    bugOccurrences[bug.bugId] = totalOccurrences;
}

int main() {
    string filename = "Test_Case_Input.tsv";
    vector<Bug> bugs = readBugsFromFile(filename);

    map<int, int> bugOccurrences;
    for (const Bug& bug : bugs) {
        if (bug.isLeaf()) {
            calculateOccurrences(bugs, bug, bugOccurrences);
        }
    }

    int maxOccurrences = 0;
    int mostAbundantBug = -1;
    for (const auto& entry : bugOccurrences) {
        int bugId = entry.first;
        int occurrences = entry.second;
        if (occurrences > maxOccurrences) {
            maxOccurrences = occurrences;
            mostAbundantBug = bugId;
        }
    }

    cout << "Most Abundant Root Bug: " << mostAbundantBug << endl;

    return 0;
}

File: online/online.py
class Bug:
    def __init__(self, edge_id, bug_id, parent_bug_id, occurrences, percentage_contribution):
        self.edge_id = edge_id
        self.bug_id = bug_id
        self.parent_bug_id = parent_bug_id
        self.occurrences = occurrences
        self.percentage_contribution = percentage_contribution

    def is_leaf(self):
        return self.occurrences is not None


def read_bugs_from_file(filename):
    bugs = []
    with open(filename, 'r') as file:
        next(file)  # Skip header line
        for line in file:
            parts = line.strip().split('\t')

            edge_id = int(parts[0])
            bug_id = int(parts[1][3:])  # Remove 'bug' prefix
            parent_bug_id = -1 if parts[2] == 'null' else int(parts[2][3:])  # Remove 'bug' prefix
            occurrences = None if parts[3] == 'null' else int(parts[3])
            percentage_contribution = 0.0 if parts[4] == 'null' else float(parts[4])

            bug = Bug(edge_id, bug_id, parent_bug_id, occurrences, percentage_contribution)
            bugs.append(bug)
    return bugs




def calculate_occurrences(bugs, bug, bug_occurrences):
    if bug.bug_id in bug_occurrences:
        return

    if bug.is_leaf():
        bug_occurrences[bug.bug_id] = bug.occurrences
        return

    total_occurrences = 0
    for child in bugs:
        if child.parent_bug_id == bug.bug_id:
            calculate_occurrences(bugs, child, bug_occurrences)
            child_occurrences = bug_occurrences[child.bug_id]
            total_occurrences += (child_occurrences * child.percentage_contribution) / 100.0

    bug_occurrences[bug.bug_id] = total_occurrences


def main():
    filename = "Test_Case_Input.tsv"
    bugs = read_bugs_from_file(filename)

    bug_occurrences = {}

    for bug in bugs:
        if bug.is_leaf():
            calculate_occurrences(bugs, bug, bug_occurrences)

    most_abundant_bug = max(bug_occurrences, key=bug_occurrences.get)
    print("Most Abundant Root Bug:", most_abundant_bug)


if __name__ == "__main__":
    main()
